### ISSUE TITLE
Feature Request: Configuration Option to always return numbers of JSON-Responses as BigDecimal

### ISSUE THREAD

**Issue by jimonthebarn** [rest-assured/rest-assured#738]
As already discussed in issue #694 I'd love to be able to configure the return type of JSON-Numbers to always be BigDecimal even for Integers. Imho this would greatly simplify verifying numbers since in most cases I really dont care if the returned number is e.g. 3.0 or 3. This is of course easy to fix if I'm in the position to change the return value of the API but that is not always the case.

I've written a small test to highlight the "issue".

`	@Test
	public void testVerificationOfNumbersOfDifferentTypes() throws Exception {

		Map<String, Object> extractedResponse = ImmutableMap.<String, Object>builder()
				.put("amount1", 1)
				.put("amount2", BigDecimal.valueOf(1))
				.build();

		//works but I need to take care that the returned value is in fact an integer
		//I'd like to highlight that the amount may be a calculated value so it is not always clear if it is an integer or a double/float value
		assertThat(extractedResponse.get("amount1"), is(1));

		//does not work due to the amount of 1 being returned as an integer instead of a BigDecimal
		assertThat(extractedResponse.get("amount1"), is(BigDecimal.ONE));

		//just works
		assertThat(extractedResponse.get("amount2"), is(BigDecimal.ONE));

	}`

Moreover the AssertionError returned by JUnit/Hamcrest does not really assist in finding the issue:
java.lang.AssertionError: 
Expected: is <1>
     but: was <1>

which doesnt really help that much.

**Comment by johanhaleby** [comment]
Hi Ben,

I understand that this would be useful. Given the way the API looks today (and I'm mainly referring to what was discussed in #694, that `NumberReturnType` is only applied to non-integers) what would be a good way forward? I don't want to break the API but I don't mind deprecating things if we can work out something good. One thing would perhaps be change `NumberReturnType` to look like this:


```java
public enum NumberReturnType {
    
    @Deprecated("use FLOATING_POINTS_AS_FLOAT_AND_DOUBLE")
    FLOAT_AND_DOUBLE,
    
    @Deprecated("use FLOATING_POINTS_AS_BIG_DECIMAL")
    BIG_DECIMAL,
    
    @Deprecated("use FLOATING_POINTS_AS_DOUBLE")
    DOUBLE,

    // Would do the same as FLOAT_AND_DOUBLE today
    FLOATING_POINTS_AS_FLOAT_AND_DOUBLE,
    
    // Would do the same as BIG_DECIMAL today
    FLOATING_POINTS_AS_BIG_DECIMAL,
    
    // Would do the same as DOUBLE today
    FLOATING_POINTS_AS_DOUBLE,

    // Treat ints and longs as they are today
    INTEGERS_AS_INTEGER_AND_LONG, 

    // Converts all int to long
    INTEGERS_AS_LONG, 

    // Convert ints and longs to BigDecimal
    INTEGERS_AS_BIG_DECIMAL, 
    
    // Convert everything to BigDecimal
    ALL_AS_BIG_DECIMAL [array of all other values in the enum]

}
```

One could then supply multiple of these values to the config (although we have to make sure that no conflicting values are supplied). Perhaps something like that could work? Any other ideas?

**Comment by jimonthebarn** [comment]
Hi Johan,

your proposal surely looks flexible enough to me. In fact I think it suffices to supply only a single of these values to the config without imposing the validation of the various possible combinations. In my opinion the most valuable feature here would be a common representation for JSON numbers. This way the resulting type would always be obvious. Then again who am I to judge how others would like to verify their numbers.

One idea that just came to mind is how about deprecating NumberReturnType altogether and instead introduce two seperate options for integer and floating point numbers e.g.:

`
public enum IntegerNumberReturnType {
    INTEGER,
    LONG,
    INTEGER_AND_LONG, 
    BIG_DECIMAL,
    
   // im not so sure about these but heck if you want youre integers to be floating point numbers why  not
    DOUBLE,
    FLOAT,
    FLOAT_AND_DOUBLE,
}

public enum FloatingNumberReturnType {
    FLOAT,
    DOUBLE
    FLOAT_AND_DOUBLE,
    BIG_DECIMAL,
}
`

By enabling to configure only a single value of each of these you could circumvent conflicting values by design and the outcome would always be well defined.

Best regards,
Ben

**Comment by johanhaleby** [comment]
I like your proposal better :) 

> You would still have to make sure that one does not configure these two new types along with the "old" NumberReturnType though.

It would probably be possible to just convert between the old `NumberReturnType` to `FloatingNumberReturnType` while it's deprecated so I don't think it should be that much of a problem.

Thanks for your feedback.

**Comment by jimonthebarn** [comment]
You're welcome. I thank you for considering. 

Cheers,
Ben

**Comment by sahajamit** [comment]
Any update on this.

**Comment by VineelaSV** [comment]
Hello folks, any update on this issue? My project requires me to have the ability to set the return type of Integers as well to BigDecimals, hence curious to know the progress on this one.

**Comment by johanhaleby** [comment]
I haven't had the time to look into it yet unfortunately, but it's something I'd like to implement. Maybe you could help out?

**Comment by dianhua1560** [comment]
Been a year, since last reply. Is this feature still worth revisit? 

IMHO, though it is easy to change the returned JSON value type while developing, it is nice to have this feature implemented.

**Comment by johanhaleby** [comment]
Yeah it would be worth revisiting I think. I think it would be a nice addition if we could implemented according to the discussions above.

**Comment by VinSV** [comment]
Yes, I think it would be a nice-to-have feature. In my case though, we
requested the developers to change the API's return value to have decimals.
For example, values were changed to "1.0" from "1".
